plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go

```diff
@@ -0,0 +1,416 @@
+		// Create subscription objects
+		subscriptions := make([]*gnmi.Subscription, len(c.Subscriptions))
+		for i, subscription := range c.Subscriptions {
+			subscriptions[i] = &gnmi.Subscription{
+				Path:              parsePath(subscription.Origin, subscription.Path, subscription.Target),
+				Mode:              gnmi.SubscriptionMode(gnmi.SubscriptionMode_value[strings.ToUpper(subscription.SubscriptionMode)]),
+				SampleInterval:    uint64(subscription.SampleInterval.Duration.Nanoseconds()),
+				SuppressRedundant: subscription.SuppressRedundant,
+				HeartbeatInterval: uint64(subscription.HeartbeatInterval.Duration.Nanoseconds()),
+			}
+		}
+
+		// Construct subscribe request
+		request := &gnmi.SubscribeRequest{
+			Request: &gnmi.SubscribeRequest_Subscribe{
+				Subscribe: &gnmi.SubscriptionList{
+					Prefix:       parsePath(c.Origin, c.Prefix, c.Target),
+					Mode:         gnmi.SubscriptionList_STREAM,
+					Encoding:     gnmi.Encoding(gnmi.Encoding_value[strings.ToUpper(c.Encoding)]),
+					Subscription: subscriptions,
+					UpdatesOnly:  c.UpdatesOnly,
+				},
+			},
+		}
+
+		subscribeClient, err := gnmi.NewGNMIClient(client).Subscribe(c.ctx)
+		if err != nil {
+			c.acc.AddError(fmt.Errorf("GNMI subscription setup failed: %v", err))
+		} else {
+			err = subscribeClient.Send(request)
+		}
+
+		if err != nil {
+			c.acc.AddError(fmt.Errorf("GNMI subscription setup failed: %v", err))
+		} else {
+			log.Printf("D! Connection to GNMI device %s established", c.ServiceAddress)
+			for c.ctx.Err() == nil {
+				reply, err := subscribeClient.Recv()
+
+				if err != nil {
+					if err != io.EOF && c.ctx.Err() == nil {
+						c.acc.AddError(fmt.Errorf("GNMI subscription aborted: %v", err))
+					}
+					break
+				}
+
+				// Check for Update message, if not skip (e.g. Sync message)
+				response, ok := reply.Response.(*gnmi.SubscribeResponse_Update)
+				if !ok {
+					continue
+				}
+
+				// Unable to derive the individual measurements or subscription without a prefix
+				if response.Update.Prefix == nil {
+					continue
+				}
+
+				timestamp := time.Unix(0, response.Update.Timestamp)
+				fields := make(map[string]interface{})
+				tags := make(map[string]string)
+
+				var builder bytes.Buffer
+
+				if len(response.Update.Prefix.Origin) > 0 {
+					builder.WriteString(response.Update.Prefix.Origin)
+					builder.WriteRune(':')
+				}
+				builder.WriteRune('/')
+
+				// Parse generic keys from prefix
+				for _, elem := range response.Update.Prefix.Elem {
+					builder.WriteString(elem.Name)
+					builder.WriteRune('/')
+
+					for key, val := range elem.Key {
+						// Use short-form of key if possible
+						if _, exists := tags[key]; exists {
+							tags[builder.String()+key] = val
+						} else {
+							tags[key] = val
+						}
+					}
+				}
+
+				tags["Producer"] = c.ServiceAddress
+				tags["Target"] = response.Update.Prefix.Target
+				builder.Truncate(builder.Len() - 1)
+				prefix := builder.String()
```

Comment:
The problem is that especially for multiple subscriptions per plugin we will receive measurements with different data structures on the same stream and there is not really a good way to differentiate other than the Prefix returned by the device.

If we were to aggregate all incoming data on the stream under the same measurement name subsequent measurements might have totally different key- or value-fields which I guess especially for InfluxDB would create a big mess.

Therefore the idea was to use the (XPath-like) path-name here to differentiate measurements since it is sort of a natural identifier for this kind of telemetry data.

Same applies more or less to the MDT plugin but for that in general it is a bit more contained environment compared to GNMI where you have different vendors on top of the existing ambiguities of the protocol.

Reply:
I'd really like to get the data into a format that is more typical to InfluxDB to it can coexist with other Telegraf metrics nicely.  To me this means having a limited set of measurement names, most plugins have 1-5, and being able to lookup the metrics and group using tags.  The measurement name should be a single word namespace for the metrics.

We may need to look over several examples.  Taking just the ones from the README I would recommend just moving the path to a tag.
```diff
- openconfig:/interfaces/interface/state/counters,name=GigabitEthernet0/0/0/0,source=10.49.234.115
+ openconfig,path=/interfaces/interface/state/counters,name=GigabitEthernet0/0/0/0,source=10.49.234.115
```

It probably gets more interesting with keys in the path though, can you show some more complex examples?